### ISSUE TITLE
Include checksum in directory listing

### ISSUE THREAD

**Issue by slithy** [xrootd/xrootd#1372]
In order to replace certain CASTOR calls with an XRootD equivalent we would like the following feature:

* Please add an option to include the checksum of each file in the output of `XrdCl::FileSystem DirList()` (`xrdfs ls`)

This is required by some of the experiments to get all the required file metadata in a single call.

**Comment by abh3** [comment]
So, I suppose we will add an option to dirlist and, if set, add the checksum. I don't want this to be the default as getting the checksum is not a trivial operation. What would you propose to happen should a file not have a checksum? Also, do you have a timeframe as to when this will be needed?

**Comment by slithy** [comment]
Thanks, Andy!

For our use case, we want to replace the equivalent call in CASTOR, which returns this struct:
```
struct Cns_filestatcs {
        u_signed64      fileid;
        mode_t          filemode;
        int             nlink;          /* number of files in a directory */
        uid_t           uid;
        gid_t           gid;
        u_signed64      filesize;
        time_t          atime;          /* last access to file */
        time_t          mtime;          /* last file modification */
        time_t          ctime;          /* last metadata modification */
        short           fileclass;
        char            status;         /* ' ' --> online, 'm' --> migrated */
        char            csumtype[3];
        char            csumvalue[CA_MAXCKSUMLEN+1];
};
```
(We don't expect an identical struct from `DirList()`, this is just for your reference).

If there is no checksum, `csumtype` will be NONE and `csumvalue` is empty.

CASTOR does not have the case of multiple checksums but EOS+CTA does. I suppose the simplest solution would be to return all the checksums as an array of (type,value) pairs.

This is required for the COMPASS experiment which will migrate to CTA around 3 months from now.

**Comment by abh3** [comment]
So, looking at what we can leverage, returning all of the checksums would be rather expensive and not necessarily complete. If we follow the current "query chksum" approach then the default is to return default chksum if a specific one was not specified. If a specific one was specified, then it must be one of the declared types. Hence, only one checksum can be returned using this approach. Returning all of them would be expensive as one would have to probe for each declared type and virtually all the of the time files in a single directory have only one of the declared types. This is due to experiments using different checksums for their files and those files don't overlap other experiment's files. So, is the the current single checksum approach sufficient (I hope so)?

**Comment by slithy** [comment]
Yes, that sounds fine. For this use case we only need ADLER32, and this approach allows the possibility of other checksums in future.

**Comment by simonmichal** [comment]
Implemented in e2f697b7b569a5de64993c80f921355294df05e5
